### DISCUSSION TITLE
Palestrante precisa ser cadastrado no banco de dados em tabela especifica?

### DISCUSSION THREAD

**Discussion by jenrodrigues** [andrebaltieri/AgendaEventos#15] [Q&A] [answered]
Ate o momento eu entendo que cada registro de evento teria um campo texto _palestrante_, no qual seria informado o nome do palestrante responsavel pela apresentacao. 
Voces acham que chegaremos ao ponto de ter uma tabela com Palestrantes pra serem relacionados aos eventos, assim como estamos fazendo com Categorias?

**Comment by LuanFreitasRibeiro** [comment]
Isso é um ponto a se discutir, temos que pensar se o palestrante é recorrente e se faz sentido mantermos o registro de palestrante. Mas também podemos pensar por outro lado: os organizadores, palestrantes e inscritos (em palestras) podem ser a mesma pessoa. 

Por exemplo: eu posso ser organizador de uma palestra, mas tbm posso me inscrever em outra e palestrar em outra, vale a pena termos uma tabela para cada um desses cenários? Ou podemos pensar em flags dentro de uma unica tabela que identifique esse cara como palestrante e organizador, caso ele não tenha a flag ele é apenas um inscrito normal?

Ficou claro o meu ponto?

**Reply by nilson-teodoro** [reply]
Pra mim ficou claro Luan, e concordo com o seu raciocínio em ter um registro do usuário e esse usuário em determinado momento pode ser, organizador, em outro momento, palestrante ou somente um participante. Acredito que com a modelagem mais madura vamos ter uma visão clara da implementação dessa solução, mas, gostei da ideia em relação a flag.

**Reply by renebentes** [reply]
Penso que podemos ter uma entidade Usuário que, poderia organizar eventos, palestrar em eventos e participar de eventos. Logo, a entidade chave seriam os eventos e a ela incluiríamos as referências conforme a função/papel que o usuário assumir perante a este? Faz sentido?

**Reply by nilson-teodoro** [reply]
Sim, concordo @renebentes

**Reply by LeonardoPedrosa** [reply]
Acredito que poderia ter uma tabela de relação entre Evento e Usuario, aonde essa tabela guardaria o idUsuario, o idEvento e o idFuncao, teríamos essa terceira tabela chamada FuncaoUsuario. ​

Exemplo:
Usuario organizador "Fulano" de Id 12 criou o Evento "Testando Evento" de Id 58.
No cadastro de palestrante ele informou que é também um dos palestrantes
Supondo que o Id da Função de organizador seja 1 e participante seja 2, a tabela EventoUsuario ficaria populada com idUsuario sendo 12, idEvento sendo 58 e idFuncao sendo 1. 
Com esse cenário o usuário logado, se for organizador, terá permissões para esse evento específico.
Sendo ele apenas um participante, ele teria seu cadastro nessa tabela com os seguintes dados, idUsuario sendo 12, idEvento = 58, idFuncao = 2 e tendo apenas permissões específicas de participante.
Penso que poderíamos ter uma função palestrante, mas para isso teríamos que cadastrar aquele palestrante em nosso sistema através do organizador, pois não faz sentido o organizador chegar no palestrante e informar que o mesmo deve acessar a plataforma para realizar seu cadastro. Acredito que podemos pensar numa melhor forma de contornar esse caso.

**Reply by renebentes** [reply]
Bom dia. Para exemplificar meu ponto de vista:

```csharp
class Usuario
{
   public string Nome { get; set; }
}

class Evento
{
   public string Titulo { get; set; }
   public Usuario Organizador { get; set; }
   public IList<Usuario> Palestrantes { get; set; }
   public IList<Usuario> Participantes { get; set; }
}
```
Faz sentido para vocês?

**Reply by LeonardoPedrosa** [reply]
Acredito que seja melhor utilizarmos ValueObject para algumas propriedades, como a propriedade "Nome", aplicamos reuso e fica mais favorável para validações. 
Para mim faz mais sentido existir realmente uma entidade chamada Palestrante, e toda vez que se for criar um evento, checar se o palestrante já está incluso no banco, se não tiver deve incluir, caso contrário apenas retorna as informações.

**Reply by JoaoNascSilva** [reply]
Creio que seja melhor separarmos em entidades distintas para uma melhor segregação e fazer uso de ValueObjects quando necessário, assim como o @LeonardoPedrosa citou. 
Acredito que venha ter mais valia desta forma.

**Reply by renebentes** [reply]
O que coloquei ali é apenas exemplificar meu ponto de vista. Também concordo com o uso de VOs. Não curto a ideia de ter várias tabelas para cada "perfil" de usuário, pois entendo que cada usuário, poderia mudar de perfil a cada evento. Em uns ele seria um organizador, outros palestrante e em outros, apenas participante.

**Comment by jenrodrigues** [comment, marked as answer]
Concluindo essa discussion porque o Balta decidiu na reuniao do dia 10 de maio que o Speaker vai ser uma entidade.
